### DISCUSSION TITLE
Multiple USB MIDI controllers from startup

### DISCUSSION THREAD

**Discussion by chestercodes** [GeekFunkLabs/fluidpatcher#12] [Q&A] [answered]
I'm hoping to use a controller for keyboard notes and then use an akai LPD8 drum pad to switch between sounds, in a similar way to described in the [Midi routing rules youtube lesson](https://youtu.be/5GmOxKsDZSg?t=457).

I've got the two controllers connected to the pi from startup and MIDI messages being received as confirmed with `aseqdump` and `aconnect`. But no sound is produced from Fluid(patcher|synth) if the keys are hit.

If I startup the pi and then connect the controllers after the LEDs blink 5 times then it works ok. Any idea why this might be? Anything I can change?

**Comment by albedozero** [comment]
Interesting - this might be the same problem another person was telling me about. My guess is that somehow two versions of FluidSynth are being started, and one is grabbing the controllers but has no soundfonts loaded. I'm traveling at the moment, but will be back home in about a week so I can test some things. What happens if you start up with the controllers connected, then unplug and re-plug them after the script starts?

**Reply by chestercodes** [reply]
Unplugging and replugging in the controllers results in them triggering sound.

**Reply by albedozero** [reply]
If you're still working on this - I've tried reproducing your issue using two MIDI controllers of my own without success. Maybe there's something different about the MIDI messages your controllers are sending - can you paste some lines of the output from `aseqdump` and `aconnect`? You might get some more feedback by running the script from the command line (after disabling the service):
```
sudo systemctl stop squishbox.service
python3 headlesspi.py
```

**Reply by chestercodes** [reply]
thanks for looking into this its very kind of you, I've played about a bit this evening, starting it up with only one controller plugged in works well and plugging the other in after means that both controllers work.

I've got it running the latest raspbian "lite" that doesn't have a GUI, i'm ssh-ing onto it to setup and run everything. 

I modified the install script to use the ADC & DAC hifiberry board and ran it from [a gist](https://gist.github.com/chestercodes/49a16f8342e48d770528d5f090c101f6)


starting it up with two controllers and a couple of other steps is shown below

```
pi@raspberrypi:~ $ aconnect -l
client 0: 'System' [type=kernel]
    0 'Timer           '
        Connecting To: 128:0
    1 'Announce        '
        Connecting To: 128:0
client 14: 'Midi Through' [type=kernel]
    0 'Midi Through Port-0'
        Connecting To: 128:0
client 16: 'LPK25' [type=kernel,card=0]
    0 'LPK25 MIDI 1    '
        Connecting To: 128:0
client 20: 'LPD8' [type=kernel,card=1]
    0 'LPD8 MIDI 1     '
        Connecting To: 128:0
client 128: 'FLUID Synth (370)' [type=user,pid=370]
    0 'Synth input port (370:0)'
        Connected From: 0:1, 14:0, 16:0, 20:0, 0:0
```

aseqdump

```
pi@raspberrypi:~ $ aseqdump -p 16
Waiting for data. Press Ctrl+C to end.
Source  Event                  Ch  Data
 16:0   Note on                 0, note 60, velocity 104
 16:0   Note off                0, note 60, velocity 127
 16:0   Note on                 0, note 62, velocity 110
 16:0   Note off                0, note 62, velocity 127
 16:0   Note on                 0, note 64, velocity 109
 16:0   Note off                0, note 64, velocity 127
 16:0   Note on                 0, note 65, velocity 108
 16:0   Note off                0, note 65, velocity 127

pi@raspberrypi:~ $ aseqdump -p 20
Waiting for data. Press Ctrl+C to end.
Source  Event                  Ch  Data
 20:0   Note on                 0, note 36, velocity 127
 20:0   Note off                0, note 36, velocity 127
 20:0   Note on                 0, note 37, velocity 127
 20:0   Note off                0, note 37, velocity 127
 20:0   Note on                 0, note 38, velocity 116
 20:0   Note off                0, note 38, velocity 127
 20:0   Note on                 0, note 39, velocity 106
 20:0   Note off                0, note 39, velocity 127

```

shutting down service and starting with headless script

```
pi@raspberrypi:~ $ sudo systemctl stop squishbox.service
pi@raspberrypi:~ $ python3 headlesspi.py
ALSA lib pcm_dmix.c:1108:(snd_pcm_dmix_open) unable to open slave
fluidsynth: error: Failed to open the "default" audio device
fluidsynth: Connection of Midi Through Port-0 succeeded
fluidsynth: Connection of LPK25 MIDI 1 succeeded
fluidsynth: Connection of LPD8 MIDI 1 succeeded
Connection is already subscribed
Connection is already subscribed
fluidsynth: warning: Sample 'SineWave': ROM sample ignored
Loading bank 'bank1.yaml' .. done!
Selected patch 1/26: Piano
```

Pulled out the cables and added both back in

```
fluidsynth: Connection of LPK25 MIDI 1 succeeded
fluidsynth: Connection of LPD8 MIDI 1 succeeded
```

Still couldn't hear anything so pulled both back out and put just keyboard back in

```
fluidsynth: Connection of LPK25 MIDI 1 succeeded
```

Played some notes and got some warnings and errors:

```
fluidsynth: warning: Failed to allocate a synthesis process. (chan=0,key=64)
fluidsynth: warning: Failed to allocate a synthesis process. (chan=0,key=67)
fluidsynth: warning: Failed to allocate a synthesis process. (chan=0,key=60)
fluidsynth: warning: Failed to allocate a synthesis process. (chan=0,key=64)
fluidsynth: warning: Ringbuffer full, try increasing synth.polyphony!
fluidsynth: warning: Ringbuffer full, try increasing synth.polyphony!
... (the above line) x 30
fluidsynth: warning: Failed to allocate a synthesis process. (chan=0,key=67)
fluidsynth: warning: Ringbuffer full, try increasing synth.polyphony!
... x 75
fluidsynth: warning: Failed to allocate a synthesis process. (chan=0,key=60)
fluidsynth: warning: Ringbuffer full, try increasing synth.polyphony!
... x 50
fluidsynth: warning: Failed to allocate a synthesis process. (chan=0,key=64)
```

stopped the python script and started it again:

```
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5047:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM default
fluidsynth: error: Failed to open the "default" audio device
fluidsynth: Connection of Midi Through Port-0 succeeded
fluidsynth: Connection of LPK25 MIDI 1 succeeded
Connection is already subscribed
fluidsynth: warning: Sample 'SineWave': ROM sample ignored
Loading bank 'bank1.yaml' .. done!
Selected patch 1/26: Piano
```

**Comment by albedozero** [comment, marked as answer]
I didn't realize you were using an add-on sound card as opposed to the headphone jack. It looks like ALSA - the Linux sound driver - is somehow mistaking your MIDI controllers for sound output devices and choosing them as the default sound device when you start up the script with one or both of them plugged in - that's why you're getting messages like:
```
ALSA lib pcm_dmix.c:1108:(snd_pcm_dmix_open) unable to open slave
fluidsynth: error: Failed to open the "default" audio device
```
and
```
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5047:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM default
fluidsynth: error: Failed to open the "default" audio device
```

When you start the script with no controllers connected, it defaults to the DAC+ ADC, since it's the only thing available. Since I don't have one of these I can't test it myself, but here are a few more suggestions to troubleshoot:

1. With the service shut down and one or both controllers plugged in, type `aplay -l` at the command line. You should just see something like:
```
**** List of PLAYBACK Hardware Devices ****
card 0: sndrpihifiberry [snd_rpi_hifiberry_dac], device 0: HifiBerry DAC HiFi pcm5102a-hifi-0 [HifiBerry DAC HiFi pcm5102a-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
i.e. just the hifiberry and no other devices. If you see your controllers there too, that at least confirms the theory that ALSA thinks they're playback devices. You could add the line `audio.alsa.device: hw:0` to `fluidsettings` in _squishboxconf.yaml_ (assuming your HiFiBerry shows up as card 0), which should force ALSA to use that card.

2. The setup file you used looks fine from your gist, but double-check _/boot/config.txt/_ to make sure the script didn't garble it somehow - maybe from running the script before your edits? To use your HiFiBerry DAC, `dtparam=audio=on` should be commented out or absent, and the only line referencing hifiberry should be `dtoverlay=hifiberry-dacplusadcpro`. Also double-check [this page](https://www.hifiberry.com/docs/software/configuring-linux-3-18-x/) to make sure you're using the right device tree overlay - are you sure you have the DAC+ ADC _Pro_ and not just the DAC+ ADC? It kind of seems like the dtoverlay is sort of working, but not quite.

3. You could try using the Pi's headphone jack for sound output, just to confirm it's an issue with the sound card/sound card setup. You can just run the setup script and choose the headphone jack in the audio setup section, and say no to everything else.

Hope one of these things works or sheds some light!

**Reply by chestercodes** [reply]
I ran the install script again and followed your suggestion in number 3 to use the headphone output. This seemed to work well so I think the problem might be related to the hifiberry.

I checked and i'm using the correct hifiberry identifier (`hifiberry-dacplusadcpro`) for the card. I guess that I might have installed it incorrectly but I wouldn't expect to hear anything at all if this was the case.
There seems to be a split in the configuration of those cards around the linux version 5.4 
I tried an [older version of raspbian](https://downloads.raspberrypi.org/raspios_lite_armhf/images/raspios_lite_armhf-2020-05-28/) that uses linux 4.19 but this hasn't fixed the problem.

One difference that I've noticed is in the output of `aplay -l` , the card number is 2 when it is started up with both controllers:

```
$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 2: sndrpihifiberry [snd_rpi_hifiberry_dacplusadcpro], device 0: HiFiBerry DAC+ADC Pro HiFi multicodec-0 [HiFiBerry DAC+ADC Pro HiFi multicodec-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

and is card 0 when there's only one controller
```
$ aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sndrpihifiberry [snd_rpi_hifiberry_dacplusadcpro], device 0: HiFiBerry DAC+ADC Pro HiFi multicodec-0 [HiFiBerry DAC+ADC Pro HiFi multicodec-0]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
```

I'm not sure why the card number and number of subdevices would change.

**Reply by albedozero** [reply]
I think this is an example of [this issue](https://support.hifiberry.com/hc/en-us/community/posts/360013083417-Designate-DAC-ADC-Pro-and-Hifibery-Microphone-Card-numbers) with the DAC+ ADC Pro - the order in which the hifiberry is detected and thus the card number it is assigned changes if other sound devices are plugged in on boot.  I feel like there must be a way to write an [asound.conf](https://www.alsa-project.org/wiki/Asoundrc#The_naming_of_PCM_devices) file that would fix this, but I haven't found it yet. For now, I think the way to get around this would be to add a bit of code at the top of _headlesspi.py_ that checks the output of `aplay -l` for the card number of the hifiberry, then passes that to fluidsynth. Give me a bit to work on that 😉

**Reply by albedozero** [reply]
OK - add these lines above the comment `# start the patcher` in _headlesspi.py_:
```
fsettings = {}
hfb = re.search("card (\d+): sndrpihifiberry", subprocess.check_output(['aplay', '-l']).decode())
if hfb: fsettings['audio.alsa.device'] = f"hw:{hfb[1]}"
```
and change the line where the Patcher is created (a couple down) to
```
    pxr = patcher.Patcher(cfgfile, fsettings)
```
...very hacky, but I think it should work.

**Reply by chestercodes** [reply]
That seems to have fixed it, thanks very much for your help!

**Reply by albedozero** [reply]
This could be a problem for other people too, so I'm glad we figured it out! Thanks for the collaboration.

**Comment by albedozero** [comment]
There's a better way - you can just add `audio.alsa.device: hw:sndrpihifiberry` to `fluidsettings` in _SquishBox/squishboxconf.yaml_. I forgot that you can also use the identifier given by `aplay -l`. not just the card number (which changes). This actually turns out to be helpful when using a USB sound card too, which I only figured out by looking at this discussion, so thanks again!
